diff --git a/wlroots-orig/xwayland/sockets.c b/wlroots/xwayland/sockets.c
index 873fde8..90eba38 100644
--- a/wlroots-orig/xwayland/sockets.c
+++ b/wlroots/xwayland/sockets.c
@@ -95,10 +95,10 @@ static bool check_socket_dir(void) {
 		wlr_log(WLR_ERROR, "%s is not a directory", socket_dir);
 		return false;
 	}
-	if (!((buf.st_uid == 0) || (buf.st_uid == getuid()))) {
-		wlr_log(WLR_ERROR, "%s not owned by root or us", socket_dir);
-		return false;
-	}
+	//if (!((buf.st_uid == 0) || (buf.st_uid == getuid()))) {
+	//	wlr_log(WLR_ERROR, "%s not owned by root or us", socket_dir);
+	//	return false;
+	//}
 	if (!(buf.st_mode & S_ISVTX)) {
 		/* we can deal with no sticky bit... */
 		if ((buf.st_mode & (S_IWGRP | S_IWOTH))) {
