From 3dbdb45218e278917f9695561f539a09f6aeaa7b Mon Sep 17 00:00:00 2001
From: Daniel Fullmer <danielrf12@gmail.com>
Date: Sun, 14 Jul 2019 17:01:18 -0400
Subject: [PATCH 3/3] Custom fingerprints

---
 .../app/attestation/server/AttestationProtocol.java   | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/src/main/java/app/attestation/server/AttestationProtocol.java b/src/main/java/app/attestation/server/AttestationProtocol.java
index 46f002b..374040d 100644
--- a/src/main/java/app/attestation/server/AttestationProtocol.java
+++ b/src/main/java/app/attestation/server/AttestationProtocol.java
@@ -165,7 +165,7 @@ class AttestationProtocol {
     private static final String ATTESTATION_APP_SIGNATURE_DIGEST_DEBUG =
             "17727D8B61D55A864936B1A7B4A2554A15151F32EBCF44CDAA6E6C3258231890";
     private static final String ATTESTATION_APP_SIGNATURE_DIGEST_RELEASE =
-            "990E04F0864B19F14F84E0E432F7A393F297AB105A22C1E1B10B442A4A62C42C";
+            "@platformFingerprint@";
     private static final int OS_VERSION_MINIMUM = 80000;
     private static final int OS_PATCH_LEVEL_MINIMUM = 201801;
     private static final int VENDOR_PATCH_LEVEL_MINIMUM = 201808;
@@ -212,6 +212,7 @@ class AttestationProtocol {
     private static final String OS_STOCK = "Stock";
     private static final String OS_GRAPHENE = "GrapheneOS";
     private static final String OS_CALYX = "CalyxOS";
+    private static final String OS_NIXDROID = "NixDroidOS";
 
     static class DeviceInfo {
         final String name;
@@ -266,6 +267,11 @@ class AttestationProtocol {
                     new DeviceInfo(DEVICE_PIXEL_2_GENERIC, 2, 3, true, true, OS_CALYX))
             .put("B4DE537A5F4B8FDAB6789EB2C06EC6E065E48A79EDD493A91F635004DD89F3E2",
                     new DeviceInfo(DEVICE_PIXEL_3_GENERIC, 3, 4, false /* uses new API */, true, OS_CALYX))
+            // NixDroidOS (TODO: support multiple devices. Needs different key for each device?)
+            .put("@taimen_avbFingerprint@",
+                    new DeviceInfo(DEVICE_PIXEL_2_GENERIC, 2, 3, true, true, OS_NIXDROID))
+            .put("@crosshatch_avbFingerprint@",
+                    new DeviceInfo(DEVICE_PIXEL_3_GENERIC, 3, 4, false /* uses new API */, true, OS_NIXDROID))
             .build();
     static final ImmutableMap<String, DeviceInfo> fingerprintsStock = ImmutableMap
             .<String, DeviceInfo>builder()
@@ -351,6 +357,9 @@ class AttestationProtocol {
             // CalyxOS
             .put("B4DE537A5F4B8FDAB6789EB2C06EC6E065E48A79EDD493A91F635004DD89F3E2",
                     new DeviceInfo(DEVICE_PIXEL_3_GENERIC, 3, 4, false /* uses new API */, true, OS_CALYX))
+            // NixDroidOS
+            .put("@crosshatch_avbFingerprint@",
+                    new DeviceInfo(DEVICE_PIXEL_3_GENERIC, 3, 4, false /* uses new API */, true, OS_NIXDROID))
             .build();
     static final ImmutableMap<String, DeviceInfo> fingerprintsStrongBoxStock = ImmutableMap
             .<String, DeviceInfo>builder()
-- 
2.22.0

