diff --git a/libbcachefs/super-io.c b/libbcachefs/super-io.c
index 482ab57..63308cf 100644
--- a/libbcachefs/super-io.c
+++ b/libbcachefs/super-io.c
@@ -1154,6 +1154,9 @@ struct replicas_status __bch2_replicas_status(struct bch_fs *c,
 				nr_offline++;
 		}
 
+                if (nr_online == 0 && nr_offline == 0)
+                    continue;
+
 		ret.replicas[e->data_type].nr_online =
 			min(ret.replicas[e->data_type].nr_online,
 			    nr_online);
